main.c

```diff
@@ -333,7 +333,7 @@ STATIC bool run_code_py(safe_mode_t safe_mode) {
     // Collects stickiness bits that apply in the current situation.
     uint8_t next_code_stickiness_situation = SUPERVISOR_NEXT_CODE_OPT_NEWLY_SET;
 
-    if (safe_mode == NO_SAFE_MODE) {
+    if (safe_mode == NO_SAFE_MODE && pyexec_mode_kind == PYEXEC_MODE_FRIENDLY_REPL) {
```

Comment:
Why put this here rather than in the loop below that calls `run_code_py`. Not calling this code at all seems clearer to me than hiding this check.

Reply:
I had similar question about `safe_mode` check -- why is it here instead of `main` function? I simply followed its example.

Would you prefer turning https://github.com/adafruit/circuitpython/blob/932131b4ff4b95066a872b5b299a84e80b7c45d3/main.c#L861

to 

```c
            if (pyexec_mode_kind == PYEXEC_MODE_FRIENDLY_REPL) {
            	skip_repl = run_code_py(safe_mode);
            } else {
            	skip_repl = false;
            }
```

instead?

This would skip the `Press any key to enter the REPL. Use CTRL-D to reload.` part, which is OK, I guess.

Reply:
Ya, that option looks better.

We have the safe mode stuff here because of the "code.py finished" state here. It used to be tightly coupled due to flashing error codes but we've removed that. We should probably refactor this to better separate the two. You don't need to do that though.

Reply:
Done now. The build was cancelled, but I think it was not because of my change.